### PR TITLE
Feat/wells

### PR DESCRIPTION
**Pull request by IvarStefansson**
Main changes:

- Geometric representation of wells 
- Discretization for coupling wells to fractures

There may be need for improvements to this functionality. Will be added as demand arises.

Wells are represented as polylines and collected in a WellNetwork, analogously to the FractureNetworks. The wells are meshed and added to an existing grid bucket. Piecewise linear grids are constructed for each well between pair of "physical" points (corresponding to a boundary, internal tip or intersection with a fracture). Meshing is done independently for each grid.

The coupling discretization is a simple implementation of the Peaceman well model.

### CODE REVIEW

File: src/porepy/numerics/ad/grid_operators.py [hunk 1]
```diff
@@ -290,35 +294,32 @@ def __init__(
...
         # requirements for processing of Ad operators.
-        self.mortar_to_primary_int = Matrix(sps.bmat([mortar_to_primary_int]).tocsr())
-        self.mortar_to_primary_avg = Matrix(sps.bmat([mortar_to_primary_avg]).tocsr())
-        self.mortar_to_secondary_int = Matrix(
-            sps.bmat([mortar_to_secondary_int]).tocsr()
-        )
-        self.mortar_to_secondary_avg = Matrix(
-            sps.bmat([mortar_to_secondary_avg]).tocsr()
-        )
+        def bmat(matrices):
```
Comment (keileg) [inline src/porepy/numerics/ad/grid_operators.py:297]: Very good.

File: src/porepy/numerics/interface_laws/elliptic_interface_laws.py [hunk 2]
```diff
@@ -851,3 +851,179 @@ def assemble_matrix_rhs(
...
+                self._primary_keyword = discr_primary.keyword
+
+        self.discr_primary = discr_primary
+        self.discr_secondary = discr_secondary
+
+        # Keys used to identify the discretization matrices of this discretization
+        self.well_discr_matrix_key = "well_mortar_discr"
+        self.mortar_vector_source_matrix_key = "well_vector_source_discr"
+
+    def ndof(self, mg: pp.MortarGrid):
```
Comment (keileg) [inline src/porepy/numerics/interface_laws/elliptic_interface_laws.py:910]:
```suggestion
    def ndof(self, mg: pp.MortarGrid) -> int:
```

File: src/porepy/numerics/interface_laws/elliptic_interface_laws.py [hunk 3]
```diff
@@ -851,3 +851,179 @@ def assemble_matrix_rhs(
...
+        return mg.num_cells
+
+    def discretize(
+        self, g_h: pp.Grid, g_l: pp.Grid, data_h: Dict, data_l: Dict, data_edge: Dict
+    ):
```
Comment (keileg) [inline src/porepy/numerics/interface_laws/elliptic_interface_laws.py:915]:
```suggestion
    )-> None:
```

File: src/porepy/numerics/interface_laws/elliptic_interface_laws.py [hunk 4]
```diff
@@ -851,3 +851,179 @@ def assemble_matrix_rhs(
...
+
+        ## Vector source.
+        # This contribution is last term of
+        # lambda = -\int{\kappa_n [p_l - p_h +  a/2 g \cdot n]} dV,
+        # where n is the outwards normal and the integral is taken over the mortar cell.
+        # (Note: This assumes a P0 discretization of mortar fluxes).
+
+        # Ambient dimension of the problem, as specified for the higher-dimensional
+        # neighbor.
+        # IMPLEMENTATION NOTE: The default value is needed to avoid that
```
Comment (keileg) [inline src/porepy/numerics/interface_laws/elliptic_interface_laws.py:973]: In my opinion, this is really an argument why vector sources by default should not be discretized. Todo.

File: src/porepy/numerics/interface_laws/elliptic_interface_laws.py [hunk 5]
```diff
@@ -851,3 +851,179 @@ def assemble_matrix_rhs(
...
+        if vector_source_dim < g_h.dim:
+            raise ValueError(
+                "Ambient dimension cannot be lower than the grid dimension"
+            )
+
+        # Construct the dot product between vectors connecting fracture and well
+        # cell centers and the identity matrix.
+
+        # Project the vectors, we need to do some transposes to get this right.
+        # Note that in the codim 2 case, proj_h maps from g_h cells, not faces.
```
Comment (keileg) [inline src/porepy/numerics/interface_laws/elliptic_interface_laws.py:997]: I was about to raise the red flag until I read this comment.

File: src/porepy/numerics/interface_laws/elliptic_interface_laws.py [hunk 6]
```diff
@@ -851,3 +851,179 @@ def assemble_matrix_rhs(
...
+        vectors_mortar = (proj_h * g_h.cell_centers.T - proj_l * g_l.cell_centers.T).T
+
+        # The values in vals are sorted by the mortar cell index ordering (proj is a
+        # csr matrix).
```
Comment (keileg) [inline src/porepy/numerics/interface_laws/elliptic_interface_laws.py:1001]: Should we do a check on that? I think this is a more general comment, we make assumptions on the matrix format here and there in the code.

File: test/unit/test_grid_bucket.py [hunk 7]
```diff
@@ -61,18 +61,18 @@ def test_add_nodes_same_grid_twice(self):
...
+        g0 = MockGrid(0)
         g1 = MockGrid(1)
-        g2 = MockGrid(2)
         g3 = MockGrid(3)
+        gb.add_nodes(g0)
         gb.add_nodes(g1)
-        gb.add_nodes(g2)
         gb.add_nodes(g3)
-        gb.add_edge([g1, g2], None)
+        gb.add_edge([g0, g1], None)
```
Comment (keileg) [inline test/unit/test_grid_bucket.py:70]:
```suggestion
        gb.add_edge([g0, g1], None)
        gb.add_edge([g1, g3], None)
```

File: src/porepy/grids/mortar_grid.py [hunk 8]
```diff
@@ -704,7 +711,11 @@ def _init_projections(
         # If the face_duplicate_ind is not given, we then assume that the primary side faces
         # already have this ordering. If the grid is created using the pp.split_grid.py
         # module this should be the case.
-        if self.num_sides() == 2 and (face_duplicate_ind is not None):
+        if (
+            self.num_sides() == 2
+            and (face_duplicate_ind is not None)
+            and self.codim < 2
```
Comment (keileg) [inline src/porepy/grids/mortar_grid.py:717]: I presume that the combination of face_duplicate_ind and codim == 2 is not meaningful. Should we explicitly give an error in that case?

File: src/porepy/fracs/wells_3d.py [hunk 9]
```diff
@@ -0,0 +1,583 @@
...
+    face_h = np.array([np.argmin(np.sum(np.power(vec, 2), axis=0))], dtype=int)
+    face_cell_map = sps.coo_matrix(
+        (np.ones(1, dtype=bool), (cell_l, face_h)), shape=(g_l.num_cells, g_h.num_faces)
+    )
+    _add_edge(g_l, g_h, gb, face_cell_map)
+
+
+def _add_edge(
+    g_l: pp.Grid, g_h: pp.Grid, gb: pp.GridBucket, primary_secondary_map: np.ndarray
+):
```
Comment (keileg) [inline src/porepy/fracs/wells_3d.py:565]:
```suggestion
) -> None:
```

File: src/porepy/fracs/wells_3d.py [hunk 10]
```diff
@@ -0,0 +1,583 @@
...
+            break  # EK, is there a preferred method?
+    cell_l = np.array([0], dtype=int)
+    cell_h = g_h.closest_cell(g_l.cell_centers)
+    cell_cell_map = sps.coo_matrix(
+        (np.ones(1, dtype=bool), (cell_l, cell_h)), shape=(g_l.num_cells, g_h.num_cells)
+    )
+    _add_edge(g_l, g_h, gb, cell_cell_map)
+
+
+def _add_well_2_intersection_edge(g_l: pp.Grid, g_h: pp.Grid, gb: pp.GridBucket):
```
Comment (keileg) [inline src/porepy/fracs/wells_3d.py:553]:
```suggestion
def _add_well_2_intersection_edge(g_l: pp.Grid, g_h: pp.Grid, gb: pp.GridBucket) -> None:
```

File: src/porepy/fracs/wells_3d.py [hunk 11]
```diff
@@ -0,0 +1,583 @@
...
+        PointGrid
+    """
+    g = pp.PointGrid(point)
+    g.name.append("Well-fracture intersection grid")
+    g.compute_geometry()
+    gb.add_nodes(g)
+    return g
+
+
+def _add_fracture_2_intersection_edge(g_l: pp.Grid, frac_num: int, gb: pp.GridBucket):
```
Comment (keileg) [inline src/porepy/fracs/wells_3d.py:536]:
```suggestion
def _add_fracture_2_intersection_edge(g_l: pp.Grid, frac_num: int, gb: pp.GridBucket) -> None:
```

File: src/porepy/fracs/wells_3d.py [hunk 12]
```diff
@@ -0,0 +1,583 @@
...
+        tol (float): Used to determine whether there is an intersection between
+            the segment and the fracture.
+
+    Returns:
+        segment_points (array, 3 x npts): updated coordinates of the points on
+            the line segment, sorted as [start, end, *any interior points].
+            Any new points have been appended.
+        tags (list of length npts): updated tags.
+    """
+    distance, isec_pt = pp.geometry.distances.segments_polygon(
```
Comment (keileg) [inline src/porepy/fracs/wells_3d.py:496]: Some comments here, please, documenting the motivation for the compuation and the main steps in the method.

File: src/porepy/fracs/wells_3d.py [hunk 13]
```diff
@@ -0,0 +1,583 @@
...
+        array (3 x npts): Sorted points.
+
+    The sorting is done so that
+    seg[d, inds[0]], seg[d,inds[1]], ..., seg[d,inds[-2]], seg[d,inds[-1]]
+    is monotone for at least one dimension d. Ascending or descending order
+    is determined by the values of the two end points.
+    """
+    for dim in range(3):
+        if not np.isclose(seg[dim, 0] - seg[dim, 1], 0):
+            break
```
Comment (keileg) [inline src/porepy/fracs/wells_3d.py:456]: The meaning of this loop is not clear.

File: src/porepy/fracs/wells_3d.py [hunk 14]
```diff
@@ -0,0 +1,583 @@
...
+                # Reset the points for next iteration/subline.
+                points_subline = np.empty((3, 0))
+                subline_endpoint_inds = [inds_seg[1]]
+                endp_tip_tags = np.zeros(2, dtype=bool)
+                endp_frac_tags = np.array([1, 0], dtype=bool)
+        for t in ["domain_boundary", "tip", "fracture"]:
+            pp.utils.tags.add_node_tags_from_face_tags(gb, t)
+
+
+def compute_well_fracture_intersections(well_network, fracture_network):
```
Comment (keileg) [inline src/porepy/fracs/wells_3d.py:382]: Typing?

**Review by keileg — COMMENTED** [review]
Mainly good. Various comments and suggestions. Let me know if you have questions (that will not be unreasonable).

File: src/porepy/fracs/wells_3d.py [hunk 15]
```diff
@@ -0,0 +1,583 @@
...
+import scipy.sparse as sps
+
+import porepy as pp
+
+# Module-wide logger
+logger = logging.getLogger(__name__)
+module_sections = ["gridding"]
+
+
+class Well(object):
```
[the remark this answers is not in the document]
Comment (IvarStefansson) [reply src/porepy/fracs/wells_3d.py:26]: The unnecessary inheritance was inherited (copied) Fracture

File: src/porepy/geometry/bounding_box.py [hunk 16]
```diff
@@ -35,3 +39,27 @@ def from_points(pts, overlap=0):
         domain["zmin"] = min_coord[2] - dx[2] * overlap
         domain["zmax"] = max_coord[2] + dx[2] * overlap
     return domain
+
+
+def make_bounding_planes_from_box(
+    box: Dict[str, float], keep_box: bool = True
```
[the remark this answers is not in the document]
Comment (IvarStefansson) [reply src/porepy/geometry/bounding_box.py:45]: I deleted it. It seems it was never used in this function.

File: src/porepy/grids/mortar_grid.py (hunk 8 above)
Comment (IvarStefansson) [reply src/porepy/grids/mortar_grid.py:717]: Done

File: src/porepy/numerics/interface_laws/elliptic_interface_laws.py (hunk 6 above)
Comment (IvarStefansson) [reply src/porepy/numerics/interface_laws/elliptic_interface_laws.py:1001]: Frankly not sure. Copied from RobinCoupling

File: src/porepy/fracs/wells_3d.py [hunk 17]
```diff
@@ -0,0 +1,583 @@
...
+        p = np.copy(self.p)
+        t = self.tags.copy()
+        return Well(p, tags=t)
+
+
+class WellNetwork3d(object):
+    """
+    Collection of Wells with geometrical information.
+
+    Facilitates meshing of all fractures in the network and their addition to
```
[the remark this answers is not in the document]
Comment (keileg) [reply src/porepy/fracs/wells_3d.py:137]:
```suggestion
    Facilitates meshing of all wells in the network and their addition to
```

File: src/porepy/fracs/wells_3d.py [hunk 18]
```diff
@@ -0,0 +1,583 @@
...
+        TODO: IS: I have not concluded whether mesh size should be global (Network)
+        or local (individual wells/segments). To be decided, with possible
+        consequences for this function.
+        """
+        size = w._mesh_size(segment_ind)
+        if size is None:
+            size = self.parameters["mesh_size"]
+        return size
+
+    def mesh(self, gb):
```
[the remark this answers is not in the document]
Comment (keileg) [reply src/porepy/fracs/wells_3d.py:228]:
```suggestion
    def mesh(self, gb: pp.GridBucket) -> None:
```

**Review by keileg — APPROVED** [review]
Two minor additions. After that, feel free to merge when you feel ready.

**Pull request merged**